### PR TITLE
Do not set auth headers for registry v1 if 302

### LINKED ISSUE #13811: BasicAuth registry login should not send Authentication header on 302
When authenticating to a registry using BasicAuth (such as a standalone registry), or when X-Docker-Token is true, we should not continue to send Authentication headers following a redirect.  This appears to be a regression introduced during 1.7 development due to RoundTripper refactoring and does not appear to affect earlier versions.

/cc @tiborvass

### PR DESCRIPTION
**Pull request by tiborvass**
This patch ensures no auth headers are set for v1 registries if there
was a 302 redirect.

This also ensures v2 does not use authTransport.

Signed-off-by: Tibor Vass tibor@docker.com

Fixes #13811

Ping @ewindisch @dmcgowan @stevvooe

### CODE REVIEW

**Comment by stevvooe** [comment]
LGTM.

File: registry/session.go [hunk 1]
```diff
@@ -70,6 +70,15 @@ func AuthTransport(base http.RoundTripper, authConfig *cliconfig.AuthConfig, alw
 }
 
 func (tr *authTransport) RoundTrip(orig *http.Request) (*http.Response, error) {
+	// Authorization should not be set on 302 redirect for untrusted locations.
+	// This logic mirrors the behavior in AddRequiredHeadersToRedirectedRequests.
+	// As the authorization logic is currently implemented in RoundTrip,
+	// a 302 redirect is detected by looking at the Referer header as go http package adds said header.
+	// This is safe as Docker doesn't set Referer in other scenarios.
+	if orig.Header.Get("Referer") != "" && !trustedLocation(orig) {
```
Comment (ewindisch) [inline registry/session.go]: The code will now only send auth headers to "trusted locations"? I guess my question is, now, if we ever want to do **that**? At least with v1 we never did this to my knowledge. The redirect was always to an unauthed location.

**Comment by dmcgowan** [comment]
LGTM

File: registry/session.go (hunk 1 above)
Comment (tiborvass) [inline registry/session.go]: @ewindisch I believe it's because of `docker.com` and `docker.io`. @dmcgowan do we ever contact `docker.com` from the code?

Comment (ewindisch) [inline registry/session.go]: @tiborvass yes, docker.com and docker.io are "trusted location" but afaik, we've never performed redirects from those to addresses which we needed to perform auth.

Comment (tiborvass) [inline registry/session.go]: Ping @dmp42

Comment (ewindisch) [inline registry/session.go]: I just checked the old code and it used to allow the redirect if the original request host *and* the destination host were trustedLocations. Still not certain if we *use* that redirect anywhere.

**Comment by calavera** [comment]
LGTM

**Comment by ewindisch** [comment]
LGTM /cc @NathanMcCauley @diogomonica

**Comment by tiborvass** [comment]
@ewindisch just to be clear I don't think we ever use the `docker.com` part of `trustedLocation` but nevermind. We can always remove it later on, but would prefer keeping it for 1.7 just to be sure.

**Comment by ewindisch** [comment]
:+1: the trustedLocation bits aren't a regression as it is, but agree we should look at it for trunk.

**Comment by icecrime** [comment]
We have 3 LGTM now, including one from a security maintainer: I'd like to move forward. Any objection @tiborvass?

**Comment by tiborvass** [comment]
@icecrime no objections.

**Comment by jfrazelle** [comment]
cherry-picked

**Comment by NathanMcCauley** [comment]
Basic question: why ever set auth headers if sending to an untrusted location?

That is, why only apply this to redirects? Shouldn't it be universal?

**Comment by ewindisch** [comment]
@NathanMcCauley "trusted location" means the DockerHub, in this context. It is entirely appropriate for private registries (or DHE) to have authentication. In that case, authentication is ONLY sent if the server supports HTTPS.

This (unshipped) *regression* appears to have allowed plaintext servers accessed using '--insecure-registry' to request and receive authentication details to be sent in the clear although this is otherwise prohibited.

Docker 1.6.2 only allowed sending auth on redirect when speaking to the DockerHub, and did not send auth on redirects for other registries. This change preserves that behavior. It is reasonable to propose denying auth-on-redirect in all cases for Engine 1.8, but that is out of scope for 1.7.

**Comment by NathanMcCauley** [comment]
@ewindisch thanks for clarification!

**Pull request merged**